mesonbuild/modules/dlang.py

```diff
@@ -16,49 +16,38 @@
 # are UI-related.
 
 import json
+from mesonbuild.interpreterbase.decorators import typed_pos_args
 import os
+import typing as T
 
 from . import ExtensionModule
 from .. import dependencies
 from .. import mlog
 from ..mesonlib import Popen_safe, MesonException
-from ..programs import ExternalProgram
+
+if T.TYPE_CHECKING:
+    from ..interpreter import Interpreter
+    from . import ModuleState
+
 
 class DlangModule(ExtensionModule):
-    class_dubbin = None
-    init_dub = False
 
-    def __init__(self, interpreter):
+    # This will always be ready since the DubDependency constructor will have
+    # been called by importing the module
+    dubbin = dependencies.DubDependency.dubbin
```

Comment:
```python
  File "/__w/meson/meson/mesonbuild/modules/dlang.py", line 37, in DlangModule
    dubbin = dependencies.DubDependency.dubbin
AttributeError: type object 'DubDependency' has no attribute 'dubbin'
```

The `__new__()` method is invoked on every instance creation, to construct a "new" copy of the class. The `__init__()` method is invoked after, to initialize the new copy of the class.

Reply:
As far as I can tell, this is only ever invoked by `dependency('foo', method: 'dub')` fully invoking `functools.partial(DubDependency, name, env, kwargs)`, which we have no guarantee comes before `import('dlang')` and certainly doesn't in our example templates. It certainly doesn't happen "by importing" anything.